src/road_gui.cpp

```diff
@@ -303,8 +303,28 @@ struct BuildRoadToolbarWindow : Window {
 	void OnInvalidateData(int data = 0, bool gui_scope = true) override
 	{
 		if (!gui_scope) return;
-
-		if (_game_mode != GM_EDITOR && !CanBuildVehicleInfrastructure(VEH_ROAD, GetRoadTramType(this->roadtype))) delete this;
+		RoadTramType rtt = GetRoadTramType(this->roadtype);
+
+		bool can_build = CanBuildVehicleInfrastructure(VEH_ROAD, rtt);
+		this->SetWidgetsDisabledState(!can_build,
+			WID_ROT_DEPOT,
+			WID_ROT_BUS_STATION,
+			WID_ROT_TRUCK_STATION,
+			WIDGET_LIST_END);
+		if (!can_build) {
+			DeleteWindowById(WC_BUS_STATION, TRANSPORT_ROAD);
+			DeleteWindowById(WC_TRUCK_STATION, TRANSPORT_ROAD);
+			DeleteWindowById(WC_BUILD_DEPOT, TRANSPORT_ROAD);
+
+			/* Show in the tooltip why this button is disabled. */
+			this->GetWidget<NWidgetCore>(WID_ROT_DEPOT)->SetDataTip(SPR_IMG_ROAD_DEPOT, STR_TOOLBAR_DISABLED_NO_VEHICLE_AVAILABLE);
+			this->GetWidget<NWidgetCore>(WID_ROT_BUS_STATION)->SetDataTip(SPR_IMG_BUS_STATION, STR_TOOLBAR_DISABLED_NO_VEHICLE_AVAILABLE);
+			this->GetWidget<NWidgetCore>(WID_ROT_TRUCK_STATION)->SetDataTip(SPR_IMG_TRUCK_BAY, STR_TOOLBAR_DISABLED_NO_VEHICLE_AVAILABLE);
+		} else {
+			this->GetWidget<NWidgetCore>(WID_ROT_DEPOT)->SetDataTip(SPR_IMG_ROAD_DEPOT, rtt == RTT_ROAD ? STR_ROAD_TOOLBAR_TOOLTIP_BUILD_ROAD_VEHICLE_DEPOT : STR_ROAD_TOOLBAR_TOOLTIP_BUILD_TRAM_VEHICLE_DEPOT);
```

Comment:
The depot sprite is roadtype specific. See SetupRoadToolbar().
Maybe store that sprite number in some member.